py/defs.bzl

```
@@ -42,9 +48,33 @@ def py_binary(name, srcs = [], main = None, imports = ["."], **kwargs):
         srcs: python source files
         main: the entry point. If absent, then the first entry in srcs is used.
         imports: List of import paths to add for this binary.
+        resolutions: FIXME
         **kwargs: additional named parameters to the py_binary_rule
     """
-    _py_binary_or_test(name = name, rule = _py_binary, srcs = srcs, main = main, imports = imports, **kwargs)
+
+    deps = kwargs.pop("deps", [])
+    concrete = []
+
+    # For a clearer DX when updating resolutions, the resolutions dict is "string" -> "label",
+    # where the rule attribute is a label-typed-dict, so reverse them here.
+    resolutions = {v: k for k, v in resolutions.items()}
+
+    for dep in deps:
+        if type(dep) == _a_struct_type:
+            if dep.virtual:
+                fail("only non-virtual deps are allowed at a py_binary or py_test rule")
+            else:
+                # constraint here must be concrete, ie == or no specifier
+                resolutions.update([["@{}_{}//:wheel".format(dep.prefix, "foo"), dep.name]])
```

This `"foo"` here looks suspicious. Shouldn't it be `dep.name` instead?

```suggestion
                resolutions.update([["@{}_{}//:wheel".format(dep.prefix, dep.name), dep.name]])
```